codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,319 @@
+
+    /** {@code streamHandler} should be {@link ChannelHandler.Sharable}. */
+    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+        this(server, streamHandler, null);
+    }
+
+    /** {@code streamHandler} should be {@link ChannelHandler.Sharable}. */
+    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler,
+            EventLoopGroup streamGroup) {
+        super(false /* selfRemoval */);
+        this.server = server;
+        this.streamHandler = streamHandler;
+        this.streamGroup = new io.netty.channel.DefaultEventLoopGroup(); //streamGroup;
+        http2Handler = new Http2ConnectionHandler(new DefaultHttp2Connection(server),
+                new Http2InboundFrameLogger(new DefaultHttp2FrameReader(), logger),
+                new Http2OutboundFrameLogger(new DefaultHttp2FrameWriter(), logger),
+                new FrameListener());
+        http2Handler.connection().addListener(new ConnectionListener());
+        add(http2Handler);
+    }
+
+    Http2ConnectionHandler connectionHandler() {
+        return http2Handler;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = ctx;
+        super.handlerAdded(ctx);
+    }
+
+    /**
+     * Handles the cleartext HTTP upgrade event. If an upgrade occurred, sends a simple response via
+     * HTTP/2 on stream 1 (the stream specifically reserved for cleartext HTTP upgrade).
+     */
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (!(evt instanceof UpgradeEvent)) {
+            super.userEventTriggered(ctx, evt);
+        }
+
+        UpgradeEvent upgrade = (UpgradeEvent) evt;
+        FullHttpRequest request = upgrade.upgradeRequest();
+        Http2Stream stream = http2Handler.connection().requireStream(1);
+        // The stream was already made active, but ctx may have been null so it wasn't initialized.
+        new ConnectionListener().onStreamActive(stream);
+        FrameListener listener = new FrameListener();
+
+        Http2Headers requestHeaders = HttpUtil.toHttp2Headers(request);
+        boolean hasContent = request.content().readableBytes() > 0;
+        boolean hasTrailers = !request.trailingHeaders().isEmpty();
+        listener.onHeadersRead(ctx, stream.id(), requestHeaders, 0,
+                !(hasContent && hasTrailers));
+        if (hasContent) {
+            listener.onDataRead(ctx, stream.id(), request.content(), 0,
+                !hasTrailers);
+        }
+        if (hasTrailers) {
+            Http2Headers headers = httpToHttp2Headers(request.trailingHeaders());
+            listener.onHeadersRead(ctx, stream.id(), headers, 0, true);
+        }
+    }
+
+    private static Http2Headers httpToHttp2Headers(HttpHeaders http1) {
+        Http2Headers http2 = new DefaultHttp2Headers();
+        for (Map.Entry<CharSequence, CharSequence> me : http1) {
```

Comment:
I don't think we need to worry about this case much. I would expect it to happen very rarely.

Note that I only wrote httpToHttp2Headers because HttpUtil didn't work for trailers and it was going to be a slight bit of work to add it (not much, but more than I wanted to do at the time). Before committing this method should be moved to HttpUtil.